src/net_processing.cpp

```diff
@@ -4286,50 +4289,56 @@ class CompareInvMempoolOrder
 };
 }
 
-bool PeerManagerImpl::SendMessages(CNode* pto)
+void PeerManagerImpl::MaybeSendPing(CNode& node_to)
 {
-    PeerRef peer = GetPeerRef(pto->GetId());
-    const Consensus::Params& consensusParams = m_chainparams.GetConsensus();
-
-    // We must call MaybeDiscourageAndDisconnect first, to ensure that we'll
-    // disconnect misbehaving peers even before the version handshake is complete.
-    if (MaybeDiscourageAndDisconnect(*pto)) return true;
-
-    // Don't send anything until the version handshake is complete
-    if (!pto->fSuccessfullyConnected || pto->fDisconnect)
-        return true;
-
-    // If we get here, the outgoing message serialization version is set and can't change.
-    const CNetMsgMaker msgMaker(pto->GetCommonVersion());
-
-    //
-    // Message: ping
-    //
+    const CNetMsgMaker msgMaker(node_to.GetCommonVersion());
     bool pingSend = false;
-    if (pto->fPingQueued) {
+
+    if (node_to.fPingQueued) {
         // RPC ping request by user
         pingSend = true;
     }
-    if (pto->nPingNonceSent == 0 && pto->m_ping_start.load() + PING_INTERVAL < GetTime<std::chrono::microseconds>()) {
+
+    if (node_to.nPingNonceSent == 0 && node_to.m_ping_start.load() + PING_INTERVAL < GetTime<std::chrono::microseconds>()) {
         // Ping automatically sent as a latency probe & keepalive.
         pingSend = true;
     }
+
     if (pingSend) {
         uint64_t nonce = 0;
         while (nonce == 0) {
             GetRandBytes((unsigned char*)&nonce, sizeof(nonce));
         }
-        pto->fPingQueued = false;
-        pto->m_ping_start = GetTime<std::chrono::microseconds>();
-        if (pto->GetCommonVersion() > BIP0031_VERSION) {
-            pto->nPingNonceSent = nonce;
-            m_connman.PushMessage(pto, msgMaker.Make(NetMsgType::PING, nonce));
+        node_to.fPingQueued = false;
+        node_to.m_ping_start = GetTime<std::chrono::microseconds>();
+        if (node_to.GetCommonVersion() > BIP0031_VERSION) {
+            node_to.nPingNonceSent = nonce;
+            m_connman.PushMessage(&node_to, msgMaker.Make(NetMsgType::PING, nonce));
         } else {
```

Comment:
tangential because you didn't change this logic, but what's the point of sending a ping message to peers who are too old to support / respond with a pong?

Reply:
pong messages were introduced in 2012 with p2p version 60000. I imagine that prior to that, the reason for a ping message was simply a keepalive (make sure that TCP connections stay up, NAT pinholes stay open, etc). This logic maintains backwards compatibility with clients older than 60000.

In practice, I expect there are almost no clients older than 60000, but if there are any it'd be a shame to arbitrarily disconnect them for failing ping timeouts.